model/src/market/proposal.rs

```diff
@@ -35,6 +39,50 @@ pub struct Proposal {
     pub prev_proposal_id: Option<String>,
 }
 
+fn choose_payment_platform(
+    demand_platforms: Vec<String>,
+    offer_platforms: Vec<String>,
+) -> Result<String, ErrorMessage> {
+    let mut platforms = HashSet::new();
+    for platform in demand_platforms {
+        platforms.insert(platform);
+    }
+
+    // Choose first matching
+    for platform in offer_platforms {
+        if platforms.contains(&platform) {
+            return Ok(platform);
+        }
+    }
+
+    Err(ErrorMessage::new("No common payment platforms".into()))
+}
+
+fn parse_payment_platforms(properties: &serde_json::value::Value) -> Vec<String> {
+    let mut platforms: Vec<String> = vec![];
+    for prop in properties.as_object().unwrap().keys() {
+        match extract_payment_platform(prop) {
+            Some(payment_platform) => platforms.push(payment_platform),
+            None => {}
+        }
+    }
+    platforms
```

Comment:
Perfect use case for `filter_map` and you can code the whole function in a single expression. :)
```suggestion
    properties.as_object().unwrap().keys().filter_map(extract_payment_platform).collect()
```

PS You might need to change the parameter type in `extract_payment_platform` from `&str` to `&String`.